internal/cmd/operator-sdk/generate/packagemanifests/packagemanifests_test.go

```diff
@@ -196,15 +196,15 @@ var _ = Describe("Running a generate packagemanifests command", func() {
 				err := c.setDefaults()
 				Expect(err).NotTo(HaveOccurred())
 				Expect(c.packageName).To(Equal("memcached-operator"))
-				Expect(c.layout).To(Equal("go.kubebuilder.io/v3"))
+				Expect(c.layout).To(Equal("go.sdk.operatorframework.io/v3"))
```

Comment:
I really think that we should use the SDK domain to let us know if the project was scaffold with SDK or KB. However,  I am not sure if we can do this change. @estroz @jmrodri @joelanford wdyt?

Reply:
I would like to point out the implications of this change:
- `operator-sdk init --plugins go/v2` still works
- The `PROJECT` file generated by the above command needs to be changed, but it already needs to be changed to remove the empty plugin configurations.
- `operator-sdk init --plugins go.kubebuilder.io/v2` will no longer work.

Conceptually, `go.sdk.kubebuilder.io/v2` is a different plugin than `go.kubebuilder.io/v2`. The first is a bundle that includes the second and three other additional plugins (`envtest/v1`, `manifests/v1` and `scorecard/v1`).

Note: `go.kubebuilder.io/v2` could always be added to the plugins of SDK but bear in mind that `go/v2` could no longer be resolved as it would be ambiguous. We would at least need `go.sdk/v2` or `go.kubebuilder/v2` to differentiate both.

Reply:
So, the all idea over integrate the projects is that SDK will scaffold the KB plugins. And then, scaffold its own customizations on top. Then:

> Conceptually, go.sdk.kubebuilder.io/v2 is a different plugin than go.kubebuilder.io/v2. 

It cannot be true.

Also, we need to allow users to choose the tool to scaffold the plugins. So, if the CLI does not find a plugin A but find the plugin B should still to do the scaffold in order to allow then take advantage of the "wrapper" plugins.

Reply:
As I already told Camila, I'm completely against scaffolding anything if some of the plugins are not found. Notifying that they are missing, sure, but using the remaining ones will leave the project in broken states too many times. Maybe not for the 2 plugins here as they may be optional and will only break its functionality, but we can't extend this requirement to every plugin.

Technically speaking, calling the bundle `"go.kubebuilder.io"` instead of `"go.sdk.operatorframework.io"` is trivial. I intentionally changed the name so that `kubebuilder init` and `operator-sdk init` result in `PROJECT` files that can be identified. Otherwise, it won't be possible to differentiate both projects but will have different scaffolded files (operator-sdk version has extra files for manifests and scorecard plugins). I think that a conversion utility from kubebuilder to SDK (changing the layout and adding the missing files) can, in this way, trivially identify if it has to convert a project or not. Either as a hook that is run transparently before every operator-sdk command, or a explicit subcommand call such as `operator-sdk import` or `operator-sdk convert`.

Reply:
> I'm completely against scaffolding anything if some of the plugins are not found. Notifying that they are missing, sure, but using the remaining ones will leave the project in broken states too many times.

:+1: 

> Technically speaking, calling the bundle "go.kubebuilder.io" instead of "go.sdk.operatorframework.io" is trivial. I intentionally changed the name so that kubebuilder init and operator-sdk init result in PROJECT files that can be identified. Otherwise, it won't be possible to differentiate both projects but will have different scaffolded files (operator-sdk version has extra files for manifests and scorecard plugins).

I don't think the projects _should_ be differentiable other than by what the default extra plugins are. That was the primary goal of migrating SDK to be kubebuilder based. 

Rather, I think the better option is to allow different binaries to add some extra default plugins to the base plugin. In the PROJECT file produced by SDK by default, you'd get something like:

```
layout: go.kubebuilder.io/v2
plugins:
  manifests.sdk.operatorframework.io/v1: {}
  scorecard.sdk.operatorframework.io/v1: {}
  envtest.sdk.operatorframework.io/v1: {}
```

And with vanilla kubebuilder, you'd get:
```
layout: go.kubebuilder.io/v2
```

Reply:
If we change the layout key, I think there will be inevitable incompatibilities where plugins will say they support ONLY go.kubebuilder.io/v2 and not go.sdk.operatorframework.io/v2, and vice versa.

And it would be easier for either project to accept drift from the other. I can foresee my future self saying "Well this is the `go.sdk.operatorframework.io/v2`" plugin, so we can make a change that's incompatible with `go.kubebuilder.io/v2` because that's a completely separate/unrelated plugin.

In the end, I don't think there should be a `go.sdk.operatorframework.io/v2` plugin. SDK is basically kubebuilder + some extra default plugins.

Reply:
@joelanford Yes, SDK is using the plugin config map to store if the additional plugins were used but this map was intended to store plugin **configuration**. For example, it has no order guarantee (it is a map, even loading and saving again this file may end up in an order change). Another example, if you run the following commands:
```
kubebuilder init
kubebuilder create api --group group --version v1 --kind KindOne --resource --controller
operator-sdk create api --group group --version v1 --kind KindTwo --resource --controller
```
it won't give any error. What should be the result? Should we have the plugin keys present with empty configs? `KindOne` doesn't have those extra files form the additional plugins but `KindTwo` does. Should we have created the files for `KindOne` (and those created at init) so that everything is consistent?

Reply:
> And it would be easier for either project to accept drift from the other. I can foresee my future self saying "Well this is the `go.sdk.operatorframework.io/v2`, plugin, so we can make a change that's incompatible with `go.kubebuilder.io/v2` because that's a completely separate/unrelated plugin."

I see your point here. This PR does a great help in this sense as there is no longer `go` plugin code in `sdk`. Not it actually uses the `kubebuilder` `go` plugin directly (before, as we couldn't chain plugins, there was a `go` plugin in SDK that basically called the `kubebuilder` one).

Reply:
+1 for the @joelanford thoughts

@Adirio, 

The problem here is not if we will or how to store the plugin config used by the tool now. So, let's keep it out of the scope and to be resolved in the future. So, I would like to fo the following proposal which in POV allows you to move forward with all your ideas and do not hurt the integration concepts and shows easily to be achieved as well.

- layout: defines the default layout of a plugin ( it could be composable as you suggested such as `base.kubebuilder.io/v1, go.kubebuilder.io/v3` or in the future `config-gen.kubebuilder.io/v1, go.kubebuilder.io/v3`). However,  it defines the core base init for the CLI.
- plugins: defines the "wrapper" plugins that were used to scaffold things on top. Some CLI can define those default plugins which will be always executed or not. So, for the SDK we would register the default plugins on the: 

```go
cli.WithDefaultCustomPlugins(
     &envtestv1.Plugin{}, // These one needs to be improved to do nothing if the project is go/v3
     &manifestsv1.Plugin{},
     &scorecardv1.Plugin{},
)

Then, the default plugins will be always be called by the CLI and they will still be stored in the plugins attr as it is today. The order that should be used is the order that we pass as arg for the CLI. 

So, I believe that the above suggestion is easily achievable, do not hurt the work that you have been doing upstream, keep the backwards compatibility, allow us to have the chains plugins and etc ... WDYT?

Reply:
> Yes, SDK is using the plugin config map to store if the additional plugins were used but this map was intended to store plugin configuration. For example, it has no order guarantee (it is a map, even loading and saving again this file may end up in an order change).

I've always seen this as something that should be fixed. IMO, the plugins used during `init` should be persisted to `PROJECT` with ordering preserved. That way other subcommands could build a pipeline of those plugins based on what's in the PROJECT file.

In your example, I would expect the SDK `create api` command NOT to use SDK's extra plugins because I wouldn't expect them to be present in the PROJECT file created by `kubebuilder init`.

Reply:
I pushed keeping the same names as in kubebuilder. It still needs to sort out how to store "extra" plugin information properly.

Reply:
Maybe now, maybe longer term... I don't really envision the need for `layout` in the PROJECT file. Just dump all of the plugins into the `plugins` field (which I think should be an ordered list). All of the plugins have the exact same interface. They take an input filesystem, and output a possibly modified filesystem (or an error).

Then the calling CLI basically does this:
```
var (
	fs Filesystem
	err error
)
for _, p := range plugins {
	fs, err = p.Run(fs)
	if err != nil {
		return err
	}
}
return fs.PersistToDisk()
```

Reply:
Okey I pushed an alternative that should keep track of the extra plugin data properly.
We are creating a bundle with exactly the same name and version as `go.kubebuilder.io/v2` that includes the other "extra" 3 plugins. These plugins store an empty `Config` object at init. `manifests` also performs some scaffolding on `create api`, so it checks if there is this config stored. If it isn't it returns `ExitError` which is a special error that means that we should not keep calling hooks of this plugin (it is returning it in pre-scaffold, so no scaffold nor post-scaffold will be called). And in the post-scaffold we update the already existing Config with the resource that this create api was called for.